File: src/ui/models/case.py
from sqlalchemy import Column, Integer, String, DateTime
from datetime import datetime
from ui.database import Base

class Case(Base):
    __tablename__ = 'cases'
    id = Column(Integer, primary_key=True)

    def __init__(self):
        pass

    def __repr__(self):
        return '<Case %r>' % (self.id)

    # General fields
    DOB = Column(DateTime)
    AgeInMonth = Column(Integer)
    Gender = Column(Integer)
    Ethnicity = Column(Integer)
    Race = Column(Integer)

    # History fields
    InjuryMech = Column(Integer)
    Amnesia_verb = Column(Integer)
    LOCSeparate = Column(Integer)
    LocLen = Column(Integer)
    Seiz = Column(Integer)
    SeizOccur = Column(Integer)
    SeizLen = Column(Integer)
    ActNorm = Column(Integer)

    # Symptoms fields
    HA_verb = Column(Integer)
    HASeverity = Column(Integer)
    HAStart = Column(Integer)
    Vomit = Column(Integer)
    VomitNbr = Column(Integer)
    VomitStart = Column(Integer)
    VomitLast = Column(Integer)
    Dizzy = Column(Integer)

    # Mental Status fields
    Intubated = Column(Integer)
    Paralyzed = Column(Integer)
    Sedated = Column(Integer)
    GCSEye = Column(Integer)
    GCSVerbal = Column(Integer)
    GCSMotor = Column(Integer)
    GCSTotal = Column(Integer)
    AMS = Column(Integer)
    AMSAgitated = Column(Integer)
    AMSSleep = Column(Integer)
    AMSSlow = Column(Integer)
    AMSRepeat = Column(Integer)
    AMSOth = Column(Integer)

    # Examination fields
    SFxPalp = Column(Integer)
    SFxPalpDepress = Column(Integer)
    FontBulg = Column(Integer)
    SFxBas = Column(Integer)
    SFxBasHem = Column(Integer)
    SFxBasRhi = Column(Integer)
    SFxBasOto = Column(Integer)
    SFxBasPer = Column(Integer)
    SFxBasRet = Column(Integer)
    Hema = Column(Integer)
    HemaLoc = Column(Integer)
    HemaSize = Column(Integer)
    Clav = Column(Integer)
    ClavFace = Column(Integer)
    ClavNeck = Column(Integer)
    ClavFro = Column(Integer)
    ClavOcc = Column(Integer)
    ClavPar = Column(Integer)
    ClavTem = Column(Integer)
    NeuroD = Column(Integer)
    NeuroDMotor = Column(Integer)
    NeuroDSensory = Column(Integer)
    NeuroDCranial = Column(Integer)
    NeuroDReflex = Column(Integer)
    NeuroDOth = Column(Integer)
    OSI = Column(Integer)
    OSIExtremity = Column(Integer)
    OSICut = Column(Integer)
    OSICspine = Column(Integer)
    OSIFlank = Column(Integer)
    OSIAbdomen = Column(Integer)
    OSIPelvis = Column(Integer)
    OSIOth = Column(Integer)
    Drugs = Column(Integer)

    # Feedback fields
    Prediction = Column(Integer)
    Evaluation = Column(Integer, nullable=True, default=None)
    Feedback = Column(String(500), nullable=True, default=None)
    DateSubmitted = Column(DateTime, nullable = True, default = datetime.utcnow())
    
    @staticmethod
    def fromForm(caseForm):
        case = Case()
        for formField in caseForm.formFields():
            for field in formField:
                if field.short_name != 'csrf_token':
                    if field.type == 'BooleanField':
                        setattr(case, field.short_name, int(field.data))
                    else:
                        setattr(case, field.short_name, field.data)
        return case

File: src/ui/static/case.js
// Javascript used in the Case view to hide fields that are not needed due to a 
// "parent" value excluding them

// update any related fields and hide or show them as needed
// this will also reset any "child" field values
function updateRelated(element, related) {
	related.forEach( function(item, idx) {
		if (element.val() > 0 || element.is(":checked")) {
			// remove the bootstrap display-none CSS
			item.removeClass('d-none')
		} else {
			// add the boostrap display-none CSS
			item.addClass('d-none')
			// set the value of all children selects to NA
			item.find('select').each( function(idx, child) {
				child.value = 92
			})
		}
	})
}

// hook the parent elements
$("#History-LOCSeparate").change(function() {
	updateRelated($(this), [$("#History-LocLen-group")])
})

$("#History-Seiz").change(function() {
	updateRelated($(this), [$("#History-SeizOccur-group")])
})

$("#Symptoms-HA_verb").change(function() {
	updateRelated($(this), [
		$("#Symptoms-HASeverity-group"),
		$("#Symptoms-HAStart-group")
	])
})

$("#Symptoms-Vomit").change(function() {
	updateRelated($(this), [
		$("#Symptoms-VomitNbr-group"),
		$("#Symptoms-VomitStart-group"),
		$("#Symptoms-VomitLast-group")
	])
})

$("#MentalStatus-AMS").change(function() {
	updateRelated($(this), [
		$("#MentalStatus-AMSSlow-group"),
		$("#MentalStatus-AMSAgitated-group"),
		$("#MentalStatus-AMSRepeat-group"),
		$("#MentalStatus-AMSSleep-group"),
		$("#MentalStatus-AMSOth-group")
	])
})

$("#Examination-SFxPalp").change(function() {
	updateRelated($(this), [
		$("#Examination-SFxPalpDepress-group"),
	])
})

$("#Examination-SFxBas").change(function() {
	updateRelated($(this), [
		$("#Examination-SFxBasHem-group"),
		$("#Examination-SFxBasRhi-group"),
		$("#Examination-SFxBasOto-group"),
		$("#Examination-SFxBasPer-group"),
		$("#Examination-SFxBasRet-group")
	])
})

$("#Examination-Hema").change(function() {
	updateRelated($(this), [
		$("#Examination-HemaLoc-group"),
		$("#Examination-HemaSize-group"),
	])
})

$("#Examination-Clav").change(function() {
	updateRelated($(this), [
		$("#Examination-ClavFace-group"),
		$("#Examination-ClavNeck-group"),
		$("#Examination-ClavFro-group"),
		$("#Examination-ClavOcc-group"),
		$("#Examination-ClavPar-group"),
		$("#Examination-ClavTem-group")
	])
})

$("#Examination-NeuroD").change(function() {
	updateRelated($(this), [
		$("#Examination-NeuroDMotor-group"),
		$("#Examination-NeuroDSensory-group"),
		$("#Examination-NeuroDCranial-group"),
		$("#Examination-NeuroDReflex-group"),
		$("#Examination-NeuroDOth-group"),
	])
})

$("#Examination-OSI").change(function() {
	updateRelated($(this), [
		$("#Examination-OSIExtremity-group"),
		$("#Examination-OSICut-group"),
		$("#Examination-OSICspine-group"),
		$("#Examination-OSIFlank-group"),
		$("#Examination-OSIAbdomen-group"),
		$("#Examination-OSIPelvis-group"),
		$("#Examination-OSIOth-group"),
	])
})

// update GCS total
function updateGCSTotal() {
	$("#MentalStatus-GCSTotal").val(parseInt($("#MentalStatus-GCSMotor").val()) 
		+ parseInt($("#MentalStatus-GCSVerbal").val())
		+ parseInt($("#MentalStatus-GCSEye").val())
	)
}
$("#MentalStatus-GCSEye").change(function() {
	updateGCSTotal()
})
$("#MentalStatus-GCSVerbal").change(function() {
	updateGCSTotal()
})
$("#MentalStatus-GCSMotor").change(function() {
	updateGCSTotal()
})

// date picker
$("#General-DOB").datetimepicker({
	format: 'D/M/YYYY',
	debug: true
})
$("#General-DOB-group").addClass("date")
$("#General-DOB").addClass("datetimepicker-input")
$("#General-DOB").attr('data-toggle', 'datetimepicker')
$("#General-DOB").attr('data-target', '#General-DOB')
$("#General-DOB").datetimepicker({
	 format: 'D/M/YYYY',
	 debug: true
})

$("#General-DOB").on("change.datetimepicker", e => {
	dob = moment($("#General-DOB").val(), 'D/M/YYYY')
	now = moment()
	$("#General-AgeInMonth").val(now.diff(dob, 'month'))
})

// set the id field as read only
$("#CaseId").attr('readonly', '')

// if the case id has a value, ensure the URL is correct
console.log($("#CaseId").val())
if ($("#CaseId").val()) {
	window.history.pushState("","", $("#CaseId").val())
}